File: ios/Runner/BLE/BlePlatformChannel.swift
import Foundation
import Flutter

class BlePlatformChannel: NSObject, FlutterStreamHandler, BLERadioCoordinatorDelegate {
    private let radioCoordinator: BLERadioCoordinator
    private var eventSink: FlutterEventSink?
    
    init(messenger: FlutterBinaryMessenger) {
        self.radioCoordinator = BLERadioCoordinator()
        super.init()
        
        self.radioCoordinator.delegate = self
        
        let methodChannel = FlutterMethodChannel(
            name: BLEConstants.controlChannelName,
            binaryMessenger: messenger
        )
        methodChannel.setMethodCallHandler(handleMethodCall)
        
        let eventChannel = FlutterEventChannel(
            name: BLEConstants.eventChannelName,
            binaryMessenger: messenger
        )
        eventChannel.setStreamHandler(self)
    }
    
    private func handleMethodCall(call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "start":
            let args = call.arguments as? [String: Any]
            let modeString = args?["mode"] as? String ?? "active"
            let mode = PowerMode(rawValue: modeString) ?? .active
            radioCoordinator.start(mode: mode)
            result(true)
            
        case "stop":
            radioCoordinator.stop()
            result(true)
            
        case "sendBroadcast":
            guard let args = call.arguments as? [String: Any],
                  let flutterData = args["data"] as? FlutterStandardTypedData else {
                result(FlutterError(code: "INVALID_ARGS", message: "Missing data payload", details: nil))
                return
            }
            radioCoordinator.sendBroadcast(data: flutterData.data)
            result(true)
            
        case "sendDirected":
            guard let args = call.arguments as? [String: Any],
                  let peerId = args["peerId"] as? String,
                  let flutterData = args["data"] as? FlutterStandardTypedData else {
                result(FlutterError(code: "INVALID_ARGS", message: "Missing peerId or data payload", details: nil))
                return
            }
            radioCoordinator.sendDirected(data: flutterData.data, targetPeerId: peerId)
            result(true)
            
        case "setPowerMode":
            guard let args = call.arguments as? [String: Any],
                  let modeString = args["mode"] as? String,
                  let mode = PowerMode(rawValue: modeString) else {
                result(FlutterError(code: "INVALID_ARGS", message: "Invalid power mode", details: nil))
                return
            }
            radioCoordinator.setPowerMode(mode)
            result(true)
            
        case "getConnectedPeers":
            result(Array(radioCoordinator.connectedPeers))
            
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    // MARK: - FlutterStreamHandler
    
    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        return nil
    }
    
    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.eventSink = nil
        return nil
    }
    
    // MARK: - BLERadioCoordinatorDelegate
    
    func radioCoordinator(_ coordinator: BLERadioCoordinator, didReceivePacket data: Data, fromPeer peerId: String, rssi: Int?) {
        var payload: [String: Any] = [
            "type": "packetReceived",
            "data": FlutterStandardTypedData(bytes: data),
            "peerId": peerId
        ]
        if let rssi = rssi {
            payload["rssi"] = rssi
        }
        eventSink?(payload)
    }
    
    func radioCoordinator(_ coordinator: BLERadioCoordinator, peerConnected peerId: String) {
        eventSink?([
            "type": "peerConnected",
            "peerId": peerId
        ])
    }
    
    func radioCoordinator(_ coordinator: BLERadioCoordinator, peerDisconnected peerId: String) {
        eventSink?([
            "type": "peerDisconnected",
            "peerId": peerId
        ])
    }
    
    func radioCoordinator(_ coordinator: BLERadioCoordinator, didUpdatePowerState isPoweredOn: Bool) {
        eventSink?([
            "type": "adapterStateChanged",
            "state": isPoweredOn ? "poweredOn" : "poweredOff"
        ])
    }
}

File: android/app/src/main/kotlin/com/bitchat/mesh/dec_chat/ble/BlePlatformChannel.kt
package com.bitchat.mesh.dec_chat.ble

import android.content.Context
import android.os.Handler
import android.os.Looper
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

class BlePlatformChannel(
    context: Context,
    messenger: BinaryMessenger
) : MethodChannel.MethodCallHandler, EventChannel.StreamHandler, BleRadioCoordinatorListener {

    private val coordinator = BleRadioCoordinator(context, this)
    private var eventSink: EventChannel.EventSink? = null
    private val mainHandler = Handler(Looper.getMainLooper())

    init {
        val methodChannel = MethodChannel(messenger, BleConstants.CONTROL_CHANNEL_NAME)
        methodChannel.setMethodCallHandler(this)

        val eventChannel = EventChannel(messenger, BleConstants.EVENT_CHANNEL_NAME)
        eventChannel.setStreamHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "start" -> {
                val modeString = call.argument<String>("mode") ?: "active"
                coordinator.start(PowerMode.fromString(modeString))
                result.success(true)
            }
            "stop" -> {
                coordinator.stop()
                result.success(true)
            }
            "sendBroadcast" -> {
                val data = call.argument<ByteArray>("data")
                if (data != null) {
                    coordinator.sendBroadcast(data)
                    result.success(true)
                } else {
                    result.error("INVALID_ARGS", "Missing data parameter", null)
                }
            }
            "sendDirected" -> {
                val peerId = call.argument<String>("peerId")
                val data = call.argument<ByteArray>("data")
                if (peerId != null && data != null) {
                    coordinator.sendDirected(peerId, data)
                    result.success(true)
                } else {
                    result.error("INVALID_ARGS", "Missing peerId or data parameter", null)
                }
            }
            "setPowerMode" -> {
                val modeString = call.argument<String>("mode") ?: "active"
                coordinator.setPowerMode(PowerMode.fromString(modeString))
                result.success(true)
            }
            "getConnectedPeers" -> {
                result.success(coordinator.getConnectedPeers())
            }
            else -> result.notImplemented()
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        this.eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        this.eventSink = null
    }

    // BleRadioCoordinatorListener
    override fun onPacketReceived(data: ByteArray, peerId: String, rssi: Int?) {
        mainHandler.post {
            val map = mutableMapOf<String, Any>(
                "type" to "packetReceived",
                "data" to data,
                "peerId" to peerId
            )
            rssi?.let { map["rssi"] = it }
            eventSink?.success(map)
        }
    }

    override fun onPeerConnected(peerId: String) {
        mainHandler.post {
            eventSink?.success(
                mapOf(
                    "type" to "peerConnected",
                    "peerId" to peerId
                )
            )
        }
    }

    override fun onPeerDisconnected(peerId: String) {
        mainHandler.post {
            eventSink?.success(
                mapOf(
                    "type" to "peerDisconnected",
                    "peerId" to peerId
                )
            )
        }
    }

    override fun onAdapterStateChanged(isPoweredOn: Boolean) {
        mainHandler.post {
            eventSink?.success(
                mapOf(
                    "type" to "adapterStateChanged",
                    "state" to if (isPoweredOn) "poweredOn" else "poweredOff"
                )
            )
        }
    }
}
